modules/standard/Lists.chpl

```diff
@@ -0,0 +1,868 @@
+    }
+
+    /*
+      Index this List via subscript. Returns a reference to the element at a
+      given index in this List.
+
+      :arg i: The index of the element to access.
+
+      :return: An element from this List.
+
+      :throws IllegalArgumentError: If the given index is out of bounds.
+    */
+    proc const this(i: int) ref throws {
+      _enter();
+
+      //
+      // TODO: Do we want to switch to array style indexing instead?
+      //
+      if false && !_boundsCheck(i) then {
+        _leave();
+        const msg = "Bad index into List: " + i:string;
+        halt(msg);
+      }
+
+      try _boundsCheckLeaveOnThrow(i);
+      ref result = _getRef(i);
+      _leave();
+      return result;
+    }
+
+    /*
+      Produce a serial iterator over the elements of this List.
+
+      :yields: A reference to one of the elements contained in this List.
+    */
+    iter these() ref {
+
+      //
+      // TODO: I'm not even sure of what the best way to WRITE a threadsafe
+      // iterator is, _let alone_ whether it should even be threadsafe
+      // (I mean, is there even a point when reference/iterator invalidation
+      // is still a thing?
+      //
+      for i in 1.._size do {
+        _enter();
+        ref result = _getRef(i);
+        _leave();
+        yield result;
+      }
+    }
+
+    /*
+      Write the contents of this List to a channel.
+
+      :arg ch: A channel to write to.
+
+      :throws SystemError: When a write to the channel fails.
+    */
+    proc writeThis(ch: channel) {
+      _enter();
+
+      //
+      // TODO TODO TODO TODO
+      //
+      // Presently, if I try to let these calls throw, then they will trigger
+      // an error in modules/standard/IO.chpl along the lines of "error: call
+      // to throwing function writeThis without throws, try, or try!", so
+      // I am not sure what the most idiomatic way to write this is.
+      //
+      try {
+        ch.write("[");
+
+        for i in 1..(_size - 1) do
+          ch.write(_getRef(i), ", ");
+
+        if _size > 0 then
+          ch.write(_getRef(_size));
+
+        ch.write("]");
+      } catch e {
+        _leave();
+      }
+
+      _leave();
+    }
+
+    /*
+      Returns `true` if this List contains zero elements.
+
+      :return: `true` if this List is empty.
+      :rtype: `bool`
+    */
+    inline proc isEmpty(): bool {
+      _enter();
+      var result = (_size == 0);
+      _leave();
+      return result;
+    }
+
+    /*
+      The current number of elements contained in this List.
+    */
+    inline proc const size {
+      _enter();
+      var result = _size;
+      _leave();
+      return result;
+    }
+
+    /*
+      Returns a new DefaultRectangular array containing a copy of each of the
+      elements contained in this List.
+
+      :return: A new DefaultRectangular array.
+    */
+    proc toArray(): [] eltType {
+      _enter();
+
+      //
+      // TODO: Do we want elements to do a full init here, or not?
```

Comment:
Yes, because the list still exists after the array is created, so init-copy should be called.